dotnet/Karmem.cs

```diff
@@ -5,48 +5,93 @@
 namespace Karmem;
 
 /// <summary>
-///     IViewer is the interface implemented by all viewers, in the
-///     generated code.
+///     IViewer is one marker interface implemented by all
+///     viewers living in the generated code.
+///
+///     The implementer MUST be on STRUCT and have one ulong field
+///     at index 0.
 /// </summary>
-public interface IViewer { public void KarmemSetPtr(ulong ptr); }
+public interface IViewer { }
 
 /// <summary>
-///     The Slice is equivalent of Span<T> but holds one array of struct Viewer. It's not possible
-///     to use Span<T> because the Viewer is one struct.
+///     The Slice is similar to Span<T>, but supports more than 2GB, assuming that the Reader is unmanaged.
+///     The Slice doesn't copy the values, instead it's just an encapsulation over the existent Reader data,
+///     beware that is unsafe to use if Reader.Dispose or Reader.Reset is called.
 ///
-///     The struct MUST have one ulong at the beginning of the struct (0 index).
+///     If T is one IViewer, the given T is expected to be one struct and MUST have one ulong at the
+///     beginning of the struct (0 index).
 /// </summary>
-public unsafe struct Slice<T> where T: IViewer
+public unsafe ref struct Slice<T>
 {
     private readonly ulong _ptr;
-    private readonly uint _length;
-    private readonly uint _stride;
+    private readonly ulong _length;
+    private readonly ulong _stride;
+    private readonly bool _isViewer;
 
     [MethodImpl(MethodImplOptions.AggressiveInlining)]
     public Slice(ulong ptr, uint length, uint stride)
     {
+        var x = new int[] { 0 };
+        var y = x.LongLength;
+ 
+
         _ptr = ptr;
-        _length = length;
-        _stride = stride;
+        _length = (ulong)length;
+        _stride = (ulong)stride;
+        _isViewer = typeof(IViewer).IsAssignableFrom(typeof(T));
```

Comment:
Seems like interface/protocols verification is not cheap, and that must be removed. The `IViewer` must die.